WearOAuth/oauth-device-grant/src/main/java/com/example/android/wearable/oauth/devicegrant/AuthDeviceGrantViewModel.kt

```diff
@@ -0,0 +1,172 @@
+        val deviceCode: String,
+        val interval: Int
+    )
+
+    /**
+     * Retrieve the information needed to verify the user. When performing this request, the server
+     * generates a user & device code pair. The user code is shown to the user and opened on the
+     * paired device. The device code is passed while polling the OAuth server.
+     */
+    private fun retrieveVerificationInfo(): Result<VerificationInfo> {
+        Log.d(TAG, "Retrieving verification info...")
+        return doPostRequest(
+            url = "https://oauth2.googleapis.com/device/code",
+            params = mapOf(
+                "client_id" to CLIENT_ID,
+                "scope" to "https://www.googleapis.com/auth/userinfo.profile"
+            )
+        ).map { responseJson ->
+            VerificationInfo(
+                verificationUri = responseJson.getString("verification_url"),
+                userCode = responseJson.getString("user_code"),
+                deviceCode = responseJson.getString("device_code"),
+                interval = responseJson.getInt("interval")
+            )
+        }
+    }
+
+    /**
+     * Opens the verification URL on the paired device.
+     *
+     * When the user has the corresponding app installed on their paired Android device, the Data
+     * Layer can be used instead, see https://developer.android.com/training/wearables/data-layer.
+     *
+     * When the user has the corresponding app installed on their paired iOS device, it should
+     * use [Universal Links](https://developer.apple.com/ios/universal-links/) to intercept the
+     * intent.
+     */
+    private fun fireRemoteIntent(verificationUri: String) {
+        remoteIntentHelper.startRemoteActivity(
+            Intent(Intent.ACTION_VIEW).apply {
+                data = Uri.parse(verificationUri)
+            },
+            null
+        )
+    }
+
+    /**
+     * Poll the Auth server for the token. This will only return when the user has finished their
+     * authorization flow on the paired device.
+     *
+     * For this sample the various exceptions aren't handled.
+     */
+    private tailrec suspend fun retrieveToken(deviceCode: String, interval: Int): String {
+        Log.d(TAG, "Polling for token...")
+        return fetchToken(deviceCode).getOrElse {
+            Log.d(TAG, "No token yet. Waiting...")
+            delay(interval * 1000L)
+            return retrieveToken(deviceCode, interval)
+        }
+    }
+
+    private fun fetchToken(deviceCode: String): Result<String> {
+        return try {
+            doPostRequest(
+                url = "https://oauth2.googleapis.com/token",
+                params = mapOf(
+                    "client_id" to CLIENT_ID,
+                    "client_secret" to CLIENT_SECRET,
+                    "device_code" to deviceCode,
+                    "grant_type" to "urn:ietf:params:oauth:grant-type:device_code"
+                )
+            ).map { responseJson ->
```

Comment:
I think `mapCatching` might be a bit nicer here than throwing an `Exception`, wrapping the block in a `try`/`catch`, and then returning a new exception.